### PR TITLE
test(otel): add GCM exporter integration test

### PR DESCRIPTION
**Pull request by dbolduc**
Part of the work for #14047

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14097)

### CODE REVIEW

File: google/cloud/opentelemetry/integration_tests/monitoring_exporter_integration_test.cc [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+  auto instrument_selector = metrics_sdk::InstrumentSelectorFactory::Create(
+      metrics_sdk::InstrumentType::kCounter, kMeterName, unit);
+
+  auto meter_selector = metrics_sdk::MeterSelectorFactory::Create(
+      kMeterName, kSchemaVersion, kSchema);
+
+  auto sum_view = metrics_sdk::ViewFactory::Create(
+      kMeterName, "description", unit, metrics_sdk::AggregationType::kSum);
+
+  p->AddView(std::move(instrument_selector), std::move(meter_selector),
```
[the remark this answers is not in the document]
Comment (dbolduc) [reply google/cloud/opentelemetry/integration_tests/monitoring_exporter_integration_test.cc]: It's not needed. Removed, thanks.

File: google/cloud/opentelemetry/integration_tests/monitoring_exporter_integration_test.cc [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+      std::move(exporter), options);
+
+  auto u_provider = metrics_sdk::MeterProviderFactory::Create(
+      std::make_unique<opentelemetry::sdk::metrics::ViewRegistry>(), resource);
+  auto* p = static_cast<metrics_sdk::MeterProvider*>(u_provider.get());
+
+  p->AddMetricReader(std::move(reader));
+
+  // counter view
+  std::string unit = "1";
```
[the remark this answers is not in the document]
Comment (dbolduc) [reply google/cloud/opentelemetry/integration_tests/monitoring_exporter_integration_test.cc]: moot: I removed the view.

https://cloud.google.com/monitoring/api/ref_v3/rest/v3/projects.metricDescriptors#MetricDescriptor.FIELDS.unit

`1` is valid, but `{iterations}` would have been more descriptive.

File: google/cloud/opentelemetry/integration_tests/monitoring_exporter_integration_test.cc [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+void DoWork(std::string const& run_id) {
+  auto provider = metrics_api::Provider::GetMeterProvider();
+  opentelemetry::nostd::shared_ptr<metrics_api::Meter> meter =
+      provider->GetMeter(kMeterName, "1.2.0");
+  auto double_counter = meter->CreateDoubleCounter(kMeterName);
+
+  // This takes 10s to run. That is unfortunate, but necessary because GCM has a
+  // minimum update period of 5s.
+  for (uint32_t i = 0; i < 20; ++i) {
```
[the remark this answers is not in the document]
Comment (dbolduc) [reply google/cloud/opentelemetry/integration_tests/monitoring_exporter_integration_test.cc]: Fixed.

File: google/cloud/opentelemetry/integration_tests/monitoring_exporter_integration_test.cc [hunk 4]
```diff
@@ -64,38 +60,29 @@ void InstallExporter(
   auto resource = opentelemetry::sdk::resource::Resource::Create(
       {{sc::kServiceNamespace, "gl-cpp"},
        // Host ID is meant to uniquely identify a VM. We don't care, though.
-       {sc::kHostId, "monitoring_exporter_integration_test"}},
-      kSchemaVersion);
+       {sc::kHostId, "monitoring_exporter_integration_test"}});
```
Comment (coryan) [inline google/cloud/opentelemetry/integration_tests/monitoring_exporter_integration_test.cc:63]: FWIW, `generic_task` would require less hacking I think.

**Review by coryan — APPROVED** [review]

File: google/cloud/opentelemetry/integration_tests/monitoring_exporter_integration_test.cc (hunk 4 above)
Comment (dbolduc) [reply google/cloud/opentelemetry/integration_tests/monitoring_exporter_integration_test.cc]: I agree that `generic_task` would be more straightforward. Thanks.

File: google/cloud/opentelemetry/integration_tests/monitoring_exporter_integration_test.cc [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+  return internal::Sample(generator, 8, "0123456789");
+}
+
+// If this code is unfamiliar, see the "simple" metrics example:
+// https://github.com/open-telemetry/opentelemetry-cpp/tree/2d077f8ec5315e0979a236554c81f621eb61f5b3/examples/metrics_simple
+void InstallExporter(std::unique_ptr<metrics_sdk::PushMetricExporter> exporter,
+                     std::string const& run_id) {
+  // GCM requires that metrics be tied to a Monitored Resource. Instead of using
+  // a GCP Resource Detector, which requires that the integration test be run
+  // from GCP, we set attributes that we know will map to a `generic_node`.
```
Comment (coryan) [inline google/cloud/opentelemetry/integration_tests/monitoring_exporter_integration_test.cc:59]: nit: s/generic_node/generic_task/

File: google/cloud/opentelemetry/integration_tests/monitoring_exporter_integration_test.cc [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+void DoWork(std::string const& run_id) {
+  auto provider = metrics_api::Provider::GetMeterProvider();
+  opentelemetry::nostd::shared_ptr<metrics_api::Meter> meter =
+      provider->GetMeter(kMeterName);
+  auto double_counter = meter->CreateDoubleCounter(kMeterName);
+
+  // This takes 10s to run. That is unfortunate, but necessary because GCM has a
+  // minimum update period of 5s.
+  for (std::uint32_t i = 0; i < 20; ++i) {
+    double_counter->Add(i, {{"run_id", run_id}});
```
Comment (coryan) [inline google/cloud/opentelemetry/integration_tests/monitoring_exporter_integration_test.cc:100]: The attribute is now redundant?

**Review by coryan — APPROVED** [review]

File: google/cloud/opentelemetry/integration_tests/monitoring_exporter_integration_test.cc (hunk 5 above)
Comment (dbolduc) [reply google/cloud/opentelemetry/integration_tests/monitoring_exporter_integration_test.cc]: Sloppy! Fixed.

File: google/cloud/opentelemetry/integration_tests/monitoring_exporter_integration_test.cc (hunk 6 above)
Comment (dbolduc) [reply google/cloud/opentelemetry/integration_tests/monitoring_exporter_integration_test.cc]: You are right. We can achieve the same thing by filtering on the `resource.task_id` label. Removed. Thanks and sorry for the churn.

**Pull request merged**